### PR TITLE
[react-native] Correct signatures for FlatList components

### PR DESCRIPTION
**Pull request by iRoachie**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://facebook.github.io/react-native/docs/flatlist.html#listemptycomponent>>

Adds correct signature for ListEmptyComponent, ListFooterComponent, ListHeaderComponent on the FlatList component.

### CODE REVIEW

File: types/react-native/index.d.ts [hunk 1]
```diff
@@ -3444,17 +3444,17 @@ export interface FlatListProperties<ItemT> extends ScrollViewProperties {
...
      * Rendered when the list is empty.
      */
-    ListEmptyComponent?: React.ComponentClass<any> | null
+    ListEmptyComponent?: React.ComponentClass<any> | React.ReactElement<any> | null
 
     /**
      * Rendered at the very end of the list.
      */
-    ListFooterComponent?: React.ComponentClass<any> | (() => React.ReactElement<any>) | null
+    ListFooterComponent?: React.ComponentClass<any> | React.ReactElement<any> | null
```
Comment (rbuckton) [inline types/react-native/index.d.ts]: According to the documentation:
> Can be a React Component Class, a render function, or a rendered element.

It looks like adding `React.ReactElement<any>` to the union is correct, but removing the `(() => React.ReactElement<any>)` from the union may be incorrect.

File: types/react-native/index.d.ts [hunk 2]
```diff
@@ -3444,17 +3444,17 @@ export interface FlatListProperties<ItemT> extends ScrollViewProperties {
     /**
      * Rendered when the list is empty.
      */
-    ListEmptyComponent?: React.ComponentClass<any> | null
+    ListEmptyComponent?: React.ComponentClass<any> | React.ReactElement<any> | null
```
Comment (rbuckton) [inline types/react-native/index.d.ts]: According to the documentation:
> Can be a React Component Class, a render function, or a rendered element.

Should this also include a "render function" in the union?

File: types/react-native/index.d.ts [hunk 3]
```diff
@@ -3444,17 +3444,17 @@ export interface FlatListProperties<ItemT> extends ScrollViewProperties {
...
      * Rendered at the very end of the list.
      */
-    ListFooterComponent?: React.ComponentClass<any> | (() => React.ReactElement<any>) | null
+    ListFooterComponent?: React.ComponentClass<any> | React.ReactElement<any> | null
 
     /**
      * Rendered at the very beginning of the list.
      */
-    ListHeaderComponent?: React.ComponentClass<any> | (() => React.ReactElement<any>) | null
+    ListHeaderComponent?: React.ComponentClass<any> | React.ReactElement<any> | null
```
Comment (rbuckton) [inline types/react-native/index.d.ts]: As above, per the documentation:
> Can be a React Component Class, a render function, or a rendered element.

Shouldn't the "render function" remain in the union?

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/react-native/index.d.ts (hunk 3 above)
Comment (iRoachie) [inline types/react-native/index.d.ts]: @rbuckton However that might be in the documentation. The actual flow typings do not carry a render function. https://github.com/facebook/react-native/blob/master/Libraries/Lists/FlatList.js#L80

File: types/react-native/index.d.ts (hunk 2 above)
Comment (huhuanming) [inline types/react-native/index.d.ts]:
```
  /**
   * Rendered when the list is empty. Can be a React Component Class, a render function, or
   * a rendered element.
   */
  ListEmptyComponent?: ?(ReactClass<any> | React.Element<any>),
```

but no `render function` in the source

File: types/react-native/index.d.ts (hunk 1 above)
Comment (huhuanming) [inline types/react-native/index.d.ts]:
```
 /**
   * Rendered at the bottom of all the items. Can be a React Component Class, a render function, or
   * a rendered element.
   */
  ListFooterComponent?: ?(ReactClass<any> | React.Element<any>),
```

Comment (huhuanming) [inline types/react-native/index.d.ts]: the document is not correctly.

**Review by huhuanming — CHANGES_REQUESTED** [review]
Add Source

**Comment by iRoachie** [comment]
Source code here https://github.com/facebook/react-native/blob/master/Libraries/Lists/FlatList.js#L80

File: types/react-native/index.d.ts (hunk 3 above)
Comment (alloy) [inline types/react-native/index.d.ts]: Seeing as the implementation tries to use the component like so `<ListEmptyComponent />`, it seems like you should be able to pass in a stateless function component and the flow types are not correct.

Comment (huhuanming) [inline types/react-native/index.d.ts]: 😄 This is true source.

**Comment by iRoachie** [comment]
@alloy @huhuanming What are we agreeing on? Adding function to the union?

**Comment by alloy** [comment]
@iRoachie Yeah, but be sure to verify the other props are used the same way, I only checked `ListEmptyComponent`.

**Comment by iRoachie** [comment]
Just did some tests and the component renders both using `<Compontent />` and `() => <Component />`

**Comment by iRoachie** [comment]
@alloy Changes made

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Comment by alloy** [comment]
Ace 👍

**Pull request merged**